gcc: only disable aligned_alloc for darwin build/host/target platforms

eliasnaur opened the pull request:
Before this change a Darwin gcc would output binaries that avoid
aligned_alloc, regardless of target platform. That's fine for Darwin
targets, but not for non-darwin targets such as pkgsCross.raspberryPi.

This change replaces the single configure flag with a flag for
each of build, host, target.

Idea by @trofi.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

trofi commented on pkgs/development/compilers/gcc/common/pre-configure.nix:
```diff
@@ -77,7 +77,14 @@ in lib.optionalString (hostPlatform.isSunOS && hostPlatform.is64bit) ''
+# Depend on targetPlatform instead, keeping non-cross gcc compatible with macOS
+# 10.12. Cross-compiling gcc output the same binaries as Linux gcc but
+# require macOS 10.15.
++ lib.optionalString (targetPlatform.isDarwin) ''
```
I think it's a bit more tricky: `aligned_alloc` is `gcc`'s builtin. As a result I expect it to be used both in `gcc`'s binary and target libraries built by `gcc`. That means you likely want `ac_cv_func_aligned_alloc=no` for `--host=darwin` (and any `--target=`) to keep the properties of `gcc` binary. And for `--target=darwin` to keep the properties of target libraries like `libgomp`.

In theory it would be nice if `gcc`'s build system provided hooks to affect only host or only target with `ac_cv_func_aligned_alloc=no`. But it's a bit tricky to get right. I think the canonical way to do it is to pass it via `STAGE{2,3}_CONFIGURE_FLAGS` to `gcc`.

Getting back to your issue you might get a reasonable outcome with `lib.optionalString (hostPlatform.isDarwin || targetPlatform.isDarwin)`.

eliasnaur replied:
> Getting back to your issue you might get a reasonable outcome with lib.optionalString (hostPlatform.isDarwin || targetPlatform.isDarwin).

So

```
lib.optionalString (hostPlatform.isDarwin || targetPlatform.isDarwin) ''
      export ac_cv_func_aligned_alloc=no
''
```

?

Isn't your new condition a super-set of the old, because of the `||`? My issue is that without this PR (or similar) cross-compiled binaries for Raspberry Pi don't reference `aligned_alloc` whereas cross-compiled binaries built on Linux do. It seems to me your suggestion strictly expands the number of configurations where `aligned_alloc` will be left out.

Did you perhaps mean `lib.optionalString (hostPlatform.isDarwin && targetPlatform.isDarwin)`, with `&&` instead of `||`?

trofi replied:
My suggestion was based on the following goals:
- when building on darwin build (either for linux or for darwin target) there should be no references to `aligned_allos)
- when building for target darwin (either on linux or darwin build) there should be no references to `aligned_allos)

Hence the `||`.

eliasnaur replied:
But I want to have darwin building for linux result in the same binaries as when linux builds for linux. That is, I want the `aligned_alloc` reference in the binaries because that's what linux building for linux have.

Do you know of a way to achieve that, without breaking darwin?

trofi replied:
Reading through `gcc/doc/install.texi` I found this bit (TIL!):

```
@subsubheading Overriding @command{configure} test results

Sometimes, it might be necessary to override the result of some
@command{configure} test, for example in order to ease porting to a new
system or work around a bug in a test.  The toplevel @command{configure}
script provides three variables for this:

@table @code

@cindex @code{build_configargs}
@item build_configargs
The contents of this variable is passed to all build @command{configure}
scripts.

@cindex @code{host_configargs}
@item host_configargs
The contents of this variable is passed to all host @command{configure}
scripts.

@cindex @code{target_configargs}
@item target_configargs
The contents of this variable is passed to all target @command{configure}
scripts.

@end table
```

Thus most robust way to get that behavior might be the (untested):

```nix
lib.optionalString (buildPlatform.isDarwin) ''
      export build_configargs=ac_cv_func_aligned_alloc=no
'' ++ lib.optionalString (hostPlatform.isDarwin) ''
      export host_configargs=ac_cv_func_aligned_alloc=no
'' ++ lib.optionalString (targetPlatform.isDarwin) ''
      export target_configargs=ac_cv_func_aligned_alloc=no
''
```

eliasnaur replied:
Thank you very much! Separate configure flags seem to be the correct fix, and achieves my goal og build paltform independent binaries. PTAL.

eliasnaur committed:
gcc: only disable aligned_alloc for darwin build/host/target platforms

Before this change a Darwin gcc would output binaries that avoid
aligned_alloc, regardless of target platform. That's fine for Darwin
targets, but not for non-darwin targets such as pkgsCross.raspberryPi.

This change replaces the single configure flag with a flag for
each of build, host, target.

Idea by @trofi.

eliasnaur commented:
Excuse my ignorance, but what is the next step in getting this merged? More approvals?

trofi commented:
Getting review from @NixOS/darwin-maintainers would be nice. Otherwise I support merging it to`staging`.

eliasnaur commented:
Time for `staging`, or wait a bit more?

veprbl reviewed (approved):
Looks like a sensible change.

eliasnaur commented:
> I guess this can only take place after [2023-05-15](https://github.com/NixOS/nixpkgs/issues/223562)?

Does the ban apply to darwin-only changes as well?

vcunat commented:
Well, perhaps we could say that `gcc` isn't a critical package on darwin (as clang is the default `stdenv.cc`), but _any_ breaking changes have cutoff today (except desktop stuff).

veprbl commented:
@eliasnaur Since this only affects cross, do you think you could massage this into a zero rebuild change?

winterqt commented:
If it only affects cross, it should be enough to just do the inverse of this: https://github.com/NixOS/nixpkgs/blob/120dd6a8032fa2dcf9197b0d49614b07b20581fe/pkgs/development/compilers/gcc/12/default.nix#L72-L74

eliasnaur commented:
If I make the PR zero-rebuild as you describe, won't it break cross-building a darwin `gcc` from linux? If so, I'd rather wait 2 weeks for this PR to get into staging. I have other mass-rebuild reproducibility PRs that need to wait for `staging` anyway.

Et7f3 commented:
> cross-building a darwin gcc from linux

> macOS/Darwin is a special case, as not the whole OS is open-source. It’s only possible to cross compile between aarch64-darwin and x86_64-darwin. aarch64-darwin support was recently added, so cross compilation is barely tested.

https://nix.dev/tutorials/cross-compilation#determining-the-host-platform-config

So I would say it is fine

eliasnaur commented:
> > cross-building a darwin gcc from linux
> 
> > macOS/Darwin is a special case, as not the whole OS is open-source. It’s only possible to cross compile between aarch64-darwin and x86_64-darwin. aarch64-darwin support was recently added, so cross compilation is barely tested.
> 
> https://nix.dev/tutorials/cross-compilation#determining-the-host-platform-config
> 
> So I would say it is fine

It may not work today, but won't we like to be able to cross-compile to darwin in the future?

If I seem overly caring about adding these special cases, it's because I've spent quite some time finding and fixing reproducibility ruining special cases already[0]. I'd rather not add more for future contributors to untangle.

[0] Apart from this PR, there are https://github.com/NixOS/nixpkgs/pull/226048, https://github.com/NixOS/nixpkgs/pull/223861, https://github.com/NixOS/nixpkgs/pull/225328, of which two trigger mass rebuilding.

eliasnaur commented:
The branch-off date has come and passed, so I assume this can be merged?

veprbl commented:
> The `staging` unrestricting date has come and passed, so I assume this can be merged?

According to #232237, you are correct.
Let's do it.
